cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -173,6 +195,84 @@ protected void addPartition(PartitionKey key, String path, boolean explorable, M
   }
 
   @Override
+  public PartitionConsumerResult consumePartitions(PartitionConsumerState partitionConsumerState) {
+    Set<Long> previousInProgress = partitionConsumerState.getVersionsToCheck();
+    Set<Long> noLongerInProgress = setDiff(previousInProgress, tx.getInProgress());
```

Comment:
since partitionConsumerState does no defensive copying, and your setDiff() modifies previousInProgress, this effectively alters the old state.